Multiple fixes to test_c10d.py.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25441 Multiple fixes to test_c10d.py.**

1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Differential Revision: [D17125824](https://our.internmc.facebook.com/intern/diff/D17125824/)

Commit:
Multiple fixes to test_c10d.py.

1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Differential Revision: [D17125824](https://our.internmc.facebook.com/intern/diff/D17125824/)

Commit:
Update on "Multiple fixes to test_c10d.py."


1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Differential Revision: [D17125824](https://our.internmc.facebook.com/intern/diff/D17125824/)

mrshenli commented:
@pritamdamania87  Do we actually need to fix static Python object destruction order problem in RPC before merging this PR? Otherwise, will this change capture that crash?

pritamdamania87 commented:
@mrshenli IIUC, the crash didn't show up in unit tests because we were using `fork` instead of `spawn`, not because the test failure was hidden due to the bug that this is fixing. @zhaojuanmao  can confirm.

zhaojuanmao commented:
yes, this diff will not catch the crash issue because it is using fork.

zhaojuanmao reviewed (approved):
this can be landed, otherwise we are hiding issues

mrshenli commented:
Hold on for a moment please. My last fix in #25419 didn't actually fix all errors, the error message does not match.

mrshenli commented:
#25465 will remove the failed test lines, and will add it back in a followup PR

mrshenli commented on test/test_c10d.py:
```diff
@@ -974,6 +976,7 @@ def _test_scatter_stress(self, inputs, fn):
             self.assertEqual(
                 torch.Tensor([iter + root]),
                 outputs[iter][root],
+                None,
```
Will it be more clear if we use `message="Mismatch in iteration %d for rank %d" % (iter, root)` here instead of setting `prec`?

pietern replied:
+1 -- this is what I'm doing in #25260 as well. Since I'll have to rebase there anyway, I'll fix this up there.

jfc4050 replied:
i've done this in #25557 as well

mrshenli commented:
#25470 is in ( Thanks! @jfc4050 ). Test errors should be gone after a rebase.

zhaojuanmao commented:
@pytorchbot rebase this please

Commit:
Update on "Multiple fixes to test_c10d.py."


1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Differential Revision: [D17125824](https://our.internmc.facebook.com/intern/diff/D17125824/)

mrshenli commented:
hey @pritamdamania87, a rebase to master is necessary to get [this line](https://github.com/pytorch/pytorch/blob/e26305ed609440b8b77a13594e7c04f2491a1b8d/test/test_c10d.py#L769) in. Otherwise, it will fail the CI.

Commit:
Update on "Multiple fixes to test_c10d.py."


1) There was a bug in https://github.com/pytorch/pytorch/pull/25012, where the
tests which needed to be skipped for return code checking was incorrect.
2) Added proper setup and teardown for the nccl_error tests.
3) Ensure AssertionError is not ignored for tests that skip return code
checking.

Differential Revision: [D17125824](https://our.internmc.facebook.com/intern/diff/D17125824/)
